docs/explanation/sessions-and-cookies.md

```diff
@@ -0,0 +1,462 @@
+
+### Using Sessions
+
+This is an example of a cookie session storage:
+
+```ts filename=app/sessions.server.ts
+import { createCookieSessionStorage } from "react-router";
+
+type SessionData = {
+  userId: string;
+};
+
+type SessionFlashData = {
+  error: string;
+};
+
+const { getSession, commitSession, destroySession } =
+  createCookieSessionStorage<SessionData, SessionFlashData>(
+    {
+      // a Cookie from `createCookie` or the CookieOptions to create one
+      cookie: {
+        name: "__session",
+
+        // all of these are optional
+        domain: "reactrouter.com",
+        // Expires can also be set (although maxAge overrides it when used in combination).
+        // Note that this method is NOT recommended as `new Date` creates only one date on each server deployment, not a dynamic date in the future!
+        //
+        // expires: new Date(Date.now() + 60_000),
+        httpOnly: true,
+        maxAge: 60,
+        path: "/",
+        sameSite: "lax",
+        secrets: ["s3cret1"],
+        secure: true,
+      },
+    }
+  );
+
+export { getSession, commitSession, destroySession };
+```
+
+We recommend setting up your session storage object in `app/sessions.server.ts` so all routes that need to access session data can import from the same spot.
+
+The input/output to a session storage object are HTTP cookies. `getSession()` retrieves the current session from the incoming request's `Cookie` header, and `commitSession()`/`destroySession()` provide the `Set-Cookie` header for the outgoing response.
+
+You'll use methods to get access to sessions in your `loader` and `action` functions.
+
+After retrieving a session with `getSession`, the returned session object has a handful of methods and properties:
+
+```tsx
+export async function action({
+  request,
+}: ActionFunctionArgs) {
+  const session = await getSession(
+    request.headers.get("Cookie")
+  );
+  session.get("foo");
+  session.has("bar");
+  // etc.
+}
+```
+
+See the [Session API][session-api] for more all the methods available on the session object.
+
+### Login form example
+
+A login form might look something like this:
+
+```tsx filename=app/routes/login.tsx lines=[4-7,12-14,16,21,25,33-35,46,51,56,61]
+import { data, redirect } from "react-router";
+import type { Route } from "./+types/login";
+
+import {
+  getSession,
+  commitSession,
+} from "../sessions.server";
+
+export async function loader({
+  request,
+}: Route.LoaderArgs) {
+  const session = await getSession(
+    request.headers.get("Cookie")
+  );
+
+  if (session.has("userId")) {
+    // Redirect to the home page if they are already signed in.
+    return redirect("/");
+  }
+
+  const data = { error: session.get("error") };
```

Comment:
This name is shadowing the `data` function.

Reply:
Whoops, thanks for that catch